src/page/page_cache.rs

```diff
@@ -359,7 +395,211 @@ impl<M> PageCache<LockFreeLog, M>
+        if !should_snapshot {
+            return;
+        }
+        self.log.flush();
+
+        let snapshot_opt = self.last_snapshot.take(SeqCst);
+        if snapshot_opt.is_none() {
+            // some other thread is snapshotting
+            warn!(
+                "snapshot skipped because previous attempt \
+                  appears not to have completed"
+            );
+            return;
+        }
+
+        let mut snapshot = snapshot_opt.unwrap();
+
+        let current_stable = self.log.stable_offset();
+
+        info!(
+            "snapshot starting from offset {} to {}",
+            snapshot.max_lid,
+            current_stable
+        );
+
+        let mut recovery = None;
+
+        for (log_id, bytes) in self.log.iter_from(snapshot.max_lid) {
+            if log_id >= current_stable {
+                // don't need to go farther
+                break;
+            }
+            if let Ok(prepend) = deserialize::<LoggedUpdate<P>>(&*bytes) {
+                match prepend.update {
+                    Update::Append(partial_pages) => {
+                        for partial_page in partial_pages {
+                            let r = self.t.recover(&partial_page);
+                            if r.is_some() {
+                                recovery = r;
+                            }
+                        }
+
+                        let mut lids = snapshot.pt.get_mut(&prepend.pid).unwrap();
+                        lids.push(log_id);
+                    }
+                    Update::Compact(partial_pages) => {
+                        for partial_page in partial_pages {
+                            let r = self.t.recover(&partial_page);
+                            if r.is_some() {
+                                recovery = r;
+                            }
+                        }
+
+                        snapshot.pt.insert(prepend.pid, vec![log_id]);
+                    }
+                    Update::Del => {
+                        snapshot.pt.remove(&prepend.pid);
+                        snapshot.free.push(prepend.pid);
+                    }
+                    Update::Alloc => {
+                        snapshot.free.retain(|&pid| pid != prepend.pid);
+                        if prepend.pid > snapshot.max_pid {
+                            snapshot.max_pid = prepend.pid;
+                        }
+                    }
+                }
+            }
+        }
+        snapshot.free.sort();
+        snapshot.free.reverse();
+        snapshot.max_lid = current_stable;
+        snapshot.recovery = recovery;
+
+        let raw_bytes = serialize(&snapshot, Infinite).unwrap();
+        let bytes = if self.config().get_use_compression() {
+            compress(&*raw_bytes, 5).unwrap()
+        } else {
+            raw_bytes
+        };
+        let crc64: [u8; 8] = unsafe { std::mem::transmute(crc64::crc64(&*bytes)) };
+
+        self.last_snapshot.swap(snapshot, SeqCst);
```

Comment:
as soon as we populate the snapshot `AtomicOption`, another thread may start a new snapshot at any time